src/compiler/emitter.ts

```diff
@@ -1106,6 +1109,213 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
+                    if (name.kind === SyntaxKind.Identifier) {
+                        var ch = (<Identifier>name).text.charAt(0);
+                        if (ch.toUpperCase() === ch) {
+                            emit(name);
+                        }
+                        else {
+                            write('"');
+                            emit(name);
+                            write('"');
+                        }
+                        return ch.toUpperCase() !== ch;
+                    }
+                    else {
+                        Debug.assert(name.kind === SyntaxKind.QualifiedName);
+                        emit(name);
+                    }
+                }
+
+                /// Emit an attribute name, which is quoted if it needs to be quoted. Because
+                /// these emit into an object literal property name, we don't need to be worried
+                /// about keywords, just non-identifier characters
+                function emitAttributeName(name: Identifier) {
+                    if (/[A-Za-z_]+[\w*]/.test(name.text)) {
+                        write('"');
+                        emit(name);
+                        write('"');
+                    }
+                    else {
+                        emit(name);
+                    }
+                }
+
+                /// Emit an name/value pair for an attribute (e.g. "x: 3")
+                function emitJsxAttribute(node: JsxAttribute) {
+                    emitAttributeName(node.name);
+                    write(': ');
+                    if (node.initializer) {
+                        emit(node.initializer);
+                    }
+                    else {
+                        write('true');
+                    }
+                }
+
+                function emitJsxElement(openingNode: JsxOpeningElement|JsxSelfClosingElement, children?: JsxChild[]) {
+                    // Call React.createElement(tag, ...
+                    emitLeadingComments(openingNode);
+                    write('React.createElement(');
+                    emitTagName(openingNode.tagName);
+                    write(', ');
+
+                    // Attribute list
+                    if (openingNode.attributes.length === 0) {
+                        // When there are no attributes, React wants 'null'
+                        write('null');
+                    }
+                    else {
+                        // Either emit one big object literal (no spread attribs), or
+                        // a call to React.__spread
+                        let attrs = openingNode.attributes;
+                        if (attrs.some(attr => attr.kind === SyntaxKind.JsxSpreadAttribute)) {
+                            write('React.__spread(');
+
+                            let haveOpenedObjectLiteral = false;
+                            for (var i = 0; i < attrs.length; i++) {

```

Comment:
Why do some of these cache the length and the others don't? I think we can just use a for-of in most cases, and a `forEach` whenever you need to do a null check